File: java/74. Search a 2D Matrix.java
// wrong O(log(m * n)) solution, cuz for loop will take O(m * n) times, but still pass.
class Solution {
  public static boolean searchMatrix(int[][] matrix, int target) {
    int[] flatten = new int[matrix.length * matrix[0].length]; // m * n

    for (int i = 0; i < matrix.length; i++) {
      for (int j = 0; j < matrix[i].length; j++) {
        flatten[i * matrix[0].length + j] = matrix[i][j];
      }
    }

    int left = 0, right = flatten.length - 1;

    while (left <= right) {
      int mid = (left + right) / 2;

      if (target > flatten[mid]) {
        left = mid + 1;
      } else if (target < flatten[mid]) {
        right = mid - 1;
      } else {
        return true;
      }
    }
    return false;
  }
}

// correct O(log(m * n)) solution.
class Solution {
    public static boolean searchMatrix(int[][] matrix, int target) {
        int left = 0, right = matrix.length * matrix[0].length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            int i = mid / matrix[0].length;
            int j = mid % matrix[0].length;
            
            if (target > matrix[i][j]) left = mid + 1;
            else if (target < matrix[i][j]) right = mid - 1;
            else return true;
        }
        return false;
    }
}

File: golang/problems/medium/74. Search a 2D Matrix.go
package medium

import "log"

// l,r
// 0 1 2 3 4 5 6 7 8 9 10 11
// 3*4 = 12
// n
// 5 -> 1, 1
// 5 / 4, 5 % 4
// 1 * n + 1
// 1 * n + 3
func searchMatrix(matrix [][]int, target int) bool {
	m := len(matrix)
	n := len(matrix[0])

	l, r := 0, (m*n)-1

	for l <= r {
		mid := l + (r-l)/2
		x, y := mid/n, mid%n
		if matrix[x][y] > target {
			r = mid - 1
		} else if matrix[x][y] == target {
			return true
		} else {
			l = mid + 1
		}
	}
	return false
}

func Test_searchMatrix() {
	matrix1 := [][]int{{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}}
	t1 := 3
	ans1 := searchMatrix(matrix1, t1)
	log.Printf("ans1: %v", ans1)
	matrix2 := [][]int{{1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 60}}
	t2 := 13
	ans2 := searchMatrix(matrix2, t2)
	log.Printf("ans2: %v", ans2)

}
